### DISCUSSION TITLE
Step 2. Understanding of Cloud Computing & Kubernetes

### DISCUSSION THREAD

**Discussion by choonho** [cloudforet-io/community#137] [Mentorship]
# Requirement

* Finished Step 1.(Week 1) Orientation and Fundamentals

# Study

* Estimated time: 2 Weeks

## Fundamentals of cloud computing and Kubernetes
* [Kubernetes Overview](https://kubernetes.io/docs/concepts/overview/)
* [Kubernetes Components](https://kubernetes.io/docs/concepts/overview/components/)
* [Learn Kubernetes Basics](https://kubernetes.io/docs/tutorials/kubernetes-basics/)

## Introduction to CI/CD (github actions and helm)
* [An Introduction to Helm](https://youtu.be/Zzwq9FmZdsU)

# Homework

> ***(optional)*** means you can do the advanced assignment if you have time.

## Create Personal Dockerhub
* https://www.docker.com/products/personal/

## Create NGINX web service
* Create github repository named as **myplugin**
* Create & Build Dockerfile with NGINX webserver
  * image name: myplugin
* create index.html like
  * the image file is at https://github.com/cloudforet-io/artwork/blob/main/illustrations/flying-wonny_christmas-version.png
![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/91d80477-91a2-42e1-a40c-cce1293e88d9)

* Upload two docker images to the Personal Docker Registry with tag 0.1 and 0.2
* (optional) Build CI using github action
  * Create .github/workflows directory for github action
  * When you push commit, create CI action and push docker image to your dockerhub automatically
  * The Docker Tag will be yyyy-mm-dd-{TIMESTAMP value}
* (optional) Build Multi Arch Docker Image
  * Support AMD64 / ARM64

## Kubernetes
* Create Kubernetes cluster (minikube is OK)

## Deploy your nginx in your kubernetes
* Create github repository as **private-registry**
* Create deployment.yaml and service.yaml file four your nginx image
* Deploy K8S Service, Deployment at "portal" namespace
* (optional) Install argoCD
   * Deploy your NGIX application using argoCD

## Submission (document)
* Your github repository URLs (myplugn, private-registry)
* Your personal dockerhub screenshot with two doker image
* Your personal dockerhub URL

# References

[Create Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
[Create Service](https://kubernetes.io/docs/concepts/services-networking/service/)

# Best Practice

| Item  | Link |
| ---    | ---   |
| myplugin | https://github.com/AnsahMohammad/myplugin |
| myplugin | https://github.com/arup1221/myplugin |
| private-registry | https://github.com/arup1221/private-registry |

**Comment by py-dev-nandini-12** [comment]
Hi choonho,
Can  you please review the project document that i have uploaded under tasks, please let me know your thoughts.
Thank you

**Reply by choonho** [reply]
Hi Nandini,
Great Job! You did very well what I intended. I will add optional homework for expert people.

**Reply by py-dev-nandini-12** [reply]
Thank you for the feedback, yes actually I am developing skills with these homeworks, I would solve to solve more problems.
Thank you

**Reply by py-dev-nandini-12** [reply]
Hi choonho,
I was trying the ci workflow using GitHub actions, but somehow the build is failing showing authentication error for logging in docker hub even after providing correct details. Can you provide some help?
Thank you

**Reply by choonho** [reply]
@py-dev-nandini-12 You can check Cloudforet CI file.

https://github.com/cloudforet-io/monitoring/blob/master/.github/workflows/dispatch_release.yaml

~~~
      - name: Set up QEMU
        uses: docker/setup-qemu-action@v2

      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v2

      - name: Login to Docker Hub
        uses: docker/login-action@v2
        with:
          username: ${{ secrets.CLOUDFORET_DOCKER_USERNAME }}
          password: ${{ secrets.CLOUDFORET_DOCKER_PASSWORD }}

      - name: Build and push to dockerhub
        uses: docker/build-push-action@v4
        with:
          context: .
          platforms: ${{ env.ARCH }}
          push: true
          tags: cloudforet/${{ github.event.repository.name }}:${{ env.VERSION }}
          build-args: |
            PACKAGE_VERSION=${{ env.PACKAGE_VERSION }}
            BRANCH_NAME=${{ env.BRANCH_NAME }}
~~~

You have to save your DOCKER login information, USERNAME, PASSWORD at
Your Github > Settings > Secrets and variables > Actions

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/42d61670-781f-40d4-83c2-ea899ae7ffb4)

**Reply by choonho** [reply]
Password is not your login password. You have to create access token.

See
https://docs.docker.com/security/for-developers/access-tokens/

**Reply by py-dev-nandini-12** [reply]
Thank you , yes I have created access token.
Let me try again, I will bother you again if I have any doubt.
Thanks 🙂

**Reply by choonho** [reply]
Check the Scope permission. You have authority as **Read & Write**

**Reply by py-dev-nandini-12** [reply]
Hi,
<img width="1680" alt="Screenshot 2023-12-14 at 12 37 56" src="https://github.com/cloudforet-io/cloudforet/assets/127988967/f47308b8-f75c-4c3e-be97-e71a0219980a">

I could create the ci action,please have a look at the docker images.

**Reply by py-dev-nandini-12** [reply]
<img width="1680" alt="Screenshot 2023-12-14 at 12 56 57" src="https://github.com/cloudforet-io/cloudforet/assets/127988967/610560b4-23e7-4652-be21-a9a44164bd66">
This is the ci action workflow file

**Reply by py-dev-nandini-12** [reply]
Below is the output after docker pull
<img width="1680" alt="Screenshot 2023-12-14 at 13 30 21" src="https://github.com/cloudforet-io/cloudforet/assets/127988967/0e2c3794-10f3-44c9-bc85-7ea9aeff11d7">

**Reply by py-dev-nandini-12** [reply]
Built  Multi Arch Docker Image
<img width="1680" alt="Screenshot 2023-12-14 at 14 17 20" src="https://github.com/cloudforet-io/cloudforet/assets/127988967/1081e59d-9a77-4e5e-945c-d7cae6d43bad">

**Reply by py-dev-nandini-12** [reply]
Hello choonho,
I was trying to deploy using argoCD but sync error is occuring, can you provide some ideas,how to get through?
Thank you

**Reply by py-dev-nandini-12** [reply]
Hi,
Could you please check the images i sent above?

**Comment by Cvr421** [comment]
Hello @choonho I have uploaded the assignment Step:2 on LFX task please review and let me know your thought .

**Reply by choonho** [reply]
@Cvr421 You did very well. For github action, please see what Cloudforet did.

https://github.com/cloudforet-io/identity/blob/master/.github/workflows/dispatch_daily_build.yaml

Instead of using bash script, use docker/build-push-action.

**Reply by Cvr421** [reply]
ok @choonho

**Reply by Cvr421** [reply]
@choonho  i have updated the github workflow Please check
https://github.com/Cvr421/myplugin/blob/main/.github/workflows/docker-ci.yml

**Reply by choonho** [reply]
CI files looks good. But checking the actions list. https://github.com/Cvr421/myplugin/actions
Action was failed. I think you did configure environment variables.

You have to save your DOCKER login information, USERNAME, PASSWORD at
Your Github > Settings > Secrets and variables > Actions

**Reply by Cvr421** [reply]
@choonho i have added the git secrets and now it is working 
https://github.com/Cvr421/myplugin/actions

![image](https://github.com/cloudforet-io/cloudforet/assets/82499697/de71914c-a5fa-4a4f-a812-60ab96518504)

**Reply by Cvr421** [reply]
@choonho please review this

**Comment by justankiit** [comment]
How to create HTML file like this https://github.com/cloudforet-io/artwork/blob/main/illustrations/flying-wonny_christmas-version.png

**Comment by justankiit** [comment]
Hey @choonho @Cvr421 , can you help me with this!!!Upload two docker images to the Personal Docker Registry with tag 0.1 and 0.2!!!!
I'm getting error!!

**Comment by justankiit** [comment]
Kindly help me in this!! we can also do a meet!!

**Reply by justankiit** [reply]
I'm also trying to create two images 
![image](https://github.com/cloudforet-io/cloudforet/assets/111564362/0a7418b0-7ca4-4ace-9281-6d23531d6e20)

**Reply by AnsahMohammad** [reply]
I think the command should be :
`docker build -t ankitstwt/myplugin:0.1 .`

**Reply by justankiit** [reply]
I will definitely try out this!!

**Comment by madguy02** [comment]
@choonho i have uploaded the document for the task given, please let me know if anything that requires change or so

**Reply by choonho** [reply]
Great. You did very well. How about creating additional tasks with github action.

**Comment by justankiit** [comment]
Which command to write for creating portal namespace in kubernetes!

**Reply by choonho** [reply]
# Create index.html

https://www.w3schools.com/html/html_intro.asp

# Embed Image
https://www.w3schools.com/html/html_images.asp

**Reply by choonho** [reply]
Creating new namespace is ***kubectl create ns {Name}***. So your command is ***kubectl create ns portal***

**Comment by arup1221** [comment]
@choonho  I have uploaded the document for the task given.

**Reply by choonho** [reply]
@arup1221 You did very well. How about create optional task, CI build with github action.

**Reply by arup1221** [reply]
Yes I will do the CI task.

**Reply by arup1221** [reply]
@choonho  I did all optional tasks.

<img width="1435" alt="Screenshot 2023-12-18 at 2 20 28 AM" src="https://github.com/cloudforet-io/cloudforet/assets/94756772/705848ea-ab61-4735-bf99-9b1f44e97d87">

**Reply by arup1221** [reply]
@choonho  can you please check the CI task. Repo link(https://github.com/arup1221/myplugin)

**Comment by AnsahMohammad** [comment]
@choonho , I didn't understand why we are creating another repo `private-registry`
Is it just to contain the `deployment.yaml` and `service.yaml` ?
Can't we do it in the main `myplugin` repo itself ?

![image](https://github.com/cloudforet-io/cloudforet/assets/85027826/eeec5251-2bb6-4477-81bf-b2f0db826183)


Is there anything that i'm missing?

**Reply by choonho** [reply]
@AnsahMohammad Good Point. You can do it in the same repository. Originally there was another content in private-repository chapter. But I changed. That's why it looks like redundancy.

**Comment by ATREAY** [comment]
Hi @choonho, I've submitted the assignment for week 2. Could you please check it out and tell me if there's anything I should fix or update?

**Reply by choonho** [reply]
@ATREAY I checked your submission. I look almost good. There is one missing point in Dockerfile. I created an issue in your github repository. Check it.

**Reply by ATREAY** [reply]
Hey @choonho, I have updated the Dockerfile can you please go through it when you have a moment?

**Comment by Ayush-Sharma410** [comment]
@choonho Sorry for being late. I have uploaded the document for step2. Can you please check and review if there are any changes needed.

**Comment by Gaurav122000** [comment]
Hey @choonho I have uploaded the document can you please check it?

**Reply by choonho** [reply]
@Gaurav122000 I created issue for your missing.

https://github.com/Gaurav122000/myplugin/issues/1

**Reply by choonho** [reply]
I also created https://github.com/Gaurav122000/private-registry/issues

**Comment by arup1221** [comment]
Hello @choonho , as you mention above I have done all the task along with the optional ones. Could I resubmit the doc?

**Reply by choonho** [reply]
@arup1221 You don't need to submit.

**Reply by arup1221** [reply]
ok, Thank You

**Comment by sarniu** [comment]
Hi @choonho, I have uploaded the document. Could you pls have a look at it and let me know if any changes or corrections needed? Thanks.

**Comment by arup1221** [comment]
Hello @choonho,  what is the next step?

**Comment by SammyBloom** [comment]
Hello @choonho ,
I'm sorry for not being online for this long. I had limited access to the internet but I'm back now.
Hope I can still make my submissions and progress in this program?
I will be submitting my Step 2 latest tomorrow.
Thank you for your understanding.

**Comment by choonho** [comment]
Hello Nwanwobi,

Good to see you again. It's good to hear from you. No worries for late
submission.
We are in Step 4. Please do your best.
Delay is not a problem.

Choonho.

2024년 1월 31일 (수) 오후 7:45, Nwanwobi Samuel ***@***.***>님이 작성:

**Comment by Suraj-kumar00** [comment]
Hi @choonho sir, i completed the task two and here I face one problem with image I tried it by giving path after uploading in vscode and also provided the direct link of the image but I don't know why it's not working but everything else is working fine.

GitHub Repo Link: https://github.com/Suraj-kumar00/myplugin
Personal DockerHub URL: https://hub.docker.com/repository/docker/surajk00/myplugin/general


GitHub Actions workflow 
![Screenshot from 2024-02-09 18-32-06](https://github.com/cloudforet-io/cloudforet/assets/123288511/c9cfde92-247a-4ba8-8b81-001d3320b7f6)

Ouput of docker pull on localhost 8080 port
![Screenshot from 2024-02-09 18-38-58](https://github.com/cloudforet-io/cloudforet/assets/123288511/1df1f99a-aad4-4463-b85f-8a53ec4632ea)

Build Multi-Arch Docker Image
![Screenshot from 2024-02-09 18-39-13](https://github.com/cloudforet-io/cloudforet/assets/123288511/19d29c1c-a78b-475d-9511-4f1f426a78de)

Deployed NGIX application using argoCD
![Screenshot from 2024-02-09 19-13-44](https://github.com/cloudforet-io/cloudforet/assets/123288511/82f451c8-7417-4fb3-a9bb-7384a97712d7)
